src/components/Navigator.js

```diff
@@ -3,74 +3,141 @@ import { Subject } from 'rxjs';
 let createNavigator = () => {
   let subject = new Subject();
 
+  let possibleMoves = [];
+  let possibleDirections = {};
+
   /**
-   * Returns possible directions from state on each level.
-   * If we are at state [1, 0] depending on the map, the
-   * result might look something like this:
-   * [
-   *    {
-   *      next: [2],
-   *      previous: [0]
-   *    },
-   *    {
-   *      next: [1, 1],
-   *      previous: false
-   *    }
-   * ]
+   * Maps motion-names to position in directions array
    *
-   * @param {number[]} state State for which to return the directions
-   * @param {*[]} elements Elements to go through
-   * @param {number} level Depth of recursion. Defaults to 0
-   * @return {[directions]} List of direcitons object
+   * @type {{
+   *  left: {level: number, direction: string},
+   *  up: {level: number, direction: string},
+   *  right: {level: number, direction: string},
+   *  down: {level: number, direction: string}
+   * }}
    */
-  let getDirections = (state, elements, level = 0) => {
-    if (level  === state.length) return [];
+  let motions = {
+    left:  { level: 0, direction: 'previous' },
+    up:    { level: 1, direction: 'previous' },
+    right: { level: 0, direction: 'next' },
+    down:  { level: 1, direction: 'next' }
+  };
 
-    let index = state[level];
-    let stateLevel = state.slice(0, level);
-    let result = [getDirectionObject(elements, index, stateLevel)];
+  /**
+   * Sets the currently possible moves and
+   * re-calculates possible directions.
+   *
+   * @param {*[]} moves
+   */
+  let setPossibleMoves = (moves) => {
+    possibleMoves = moves;
+    calculatePossibleDirections();
+  };
 
-    return result.concat(getDirections(state, elements[index].children, level + 1));
+  /**
+   * Calculates possible directions based on
+   * possibleMoves
+   */
+  let calculatePossibleDirections = () => {
+    let reduce  = (result, name) => {
```

Comment:
It calculates the possible directions based on `possibleMoves`. This way we can easily check if we can go into a direction or not.